### ISSUE TITLE
Installing `--HEAD git` doesn't pick up tags and therefore provides confusing version info

### ISSUE THREAD

**Issue by me-and** [Homebrew/homebrew-core#44401]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, Homebrew/homebrew-core formula (not cask) on macOS? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a Linux problem please file this issue at https://github.com/Homebrew/linuxbrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

TL;DR: when installing Git using `brew install --HEAD git`, and doing subsequent updates/reinstalls, the repository that gets used at `~/Library/Caches/Homebrew/git--git` gets the Git config option `remote.origin.tagOpt` set to `--no-tags`, which means running `git --version` produces unexpected results.

If I install Git using `brew install --HEAD git`, Homebrew checks out the Git repository at `~/Library/Caches/Homebrew/git--git`. When Git builds, it gets its current version number from doing `git describe` (or something very similar), which is based off the Git tags in the repository.  During the initial build, this is absolutely fine: a fresh build for me currently reports `git version 2.23.0.256.g4c86140027`, which is pretty much exactly what I expect.

However, subsequent updates, reinstalls, etc, appear to not pull tags from the upstream repository, which means the call to `git describe` during the Git build process can't find the most recent tags, and the first parts of the version number reported by `git --version` don't get updated.

Slightly contrived example, because it's difficult to demonstrate when I don't have control of tagging in the upstream repository:

```
✔ adam@wheatley ~
Wed 18 Sep 21:35 $ brew uninstall git
Uninstalling /usr/local/Cellar/git/HEAD-4c86140_1... (1,544 files, 45.0MB)

✔ adam@wheatley ~
Wed 18 Sep 21:36 $ rm -rf ~/Library/Caches/Homebrew/git--git  # Clear state

✔ adam@wheatley ~
Wed 18 Sep 21:36 $ brew install --HEAD git
==> Cloning https://github.com/git/git.git
Cloning into '/Users/adam/Library/Caches/Homebrew/git--git'...
Checking out files: 100% (3645/3645), done.
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
Submodule 'sha1collisiondetection' (https://github.com/cr-marcstevens/sha1collisiondetection.git) registered for path 'sha1collisiondetection'
Cloning into '/Users/adam/Library/Caches/Homebrew/git--git/sha1collisiondetection'...
Submodule path 'sha1collisiondetection': checked out '855827c583bc30645ba427885caa40c5b81764d2'
/Users/adam/Library/Caches/Homebrew/git--git/sha1collisiondetection
==> make install prefix=/usr/local/Cellar/git/HEAD-4c86140_1 sysconfdir=/usr/local/etc CC=clang CFLAGS= LDFLAGS= NO_OPENSSL=1 APPLE_COMMON_CRYPTO=1
==> make CC=clang CFLAGS= LDFLAGS=
==> make clean
==> make
==> make test
==> make CC=clang CFLAGS= LDFLAGS=
==> Downloading https://www.kernel.org/pub/software/scm/git/git-manpages-2.23.0.tar.xz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/6a8411d0ce8a37af7bcc0b586a2e4bc4f286f54bdfdcbc2683fe76ed72ac457b--git-manpages-2.23.0.tar.xz
==> Downloading https://www.kernel.org/pub/software/scm/git/git-htmldocs-2.23.0.tar.xz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/1d663eba035cf8f0470de76ca1532a69932cb7003300e62d670ed7879d6883d4--git-htmldocs-2.23.0.tar.xz
==> Downloading https://cpan.metacpan.org/authors/id/R/RJ/RJBS/Net-SMTP-SSL-1.04.tar.gz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/0726c68dac390b617cefcc880f076b50795463c1fba286d46a53082ebed2bc9d--Net-SMTP-SSL-1.04.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d

zsh completions and functions have been installed to:
  /usr/local/share/zsh/site-functions

Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/git
==> Summary
🍺  /usr/local/Cellar/git/HEAD-4c86140_1: 1,544 files, 45.0MB, built in 3 minutes 17 seconds

✔ adam@wheatley ~
Wed 18 Sep 21:39 $ /usr/local/bin/git --version # This produces the expected version number
git version 2.23.0.256.g4c86140027

✔ adam@wheatley ~
Wed 18 Sep 21:39 $ cd ~/Library/Caches/Homebrew/git--git

✔ adam@wheatley ~/Library/Caches/Homebrew/git--git [master|✔]
Wed 18 Sep 21:40 $ git config --get remote.origin.tagOpt

✘-1 adam@wheatley ~/Library/Caches/Homebrew/git--git [master|✔]
Wed 18 Sep 21:40 $ git tag --delete v2.23.0  # Act like this tag doesn't exist
Deleted tag 'v2.23.0' (was cb71568594)

✔ adam@wheatley ~/Library/Caches/Homebrew/git--git [master|✔]
Wed 18 Sep 21:40 $ cd

✔ adam@wheatley ~
Wed 18 Sep 21:40 $ brew reinstall git  # This should produce something exactly the same, right?
==> Reinstalling git
==> Cloning https://github.com/git/git.git
Updating /Users/adam/Library/Caches/Homebrew/git--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at 4c86140027 Third batch
Entering 'sha1collisiondetection'
/Users/adam/Library/Caches/Homebrew/git--git/sha1collisiondetection
==> make install prefix=/usr/local/Cellar/git/HEAD-4c86140_1 sysconfdir=/usr/local/etc CC=clang CFLAGS= LDFLAGS= NO_OPENSSL=1 APPLE_COMMON_CRYPTO=1
==> make CC=clang CFLAGS= LDFLAGS=
==> make clean
==> make
==> make test
==> make CC=clang CFLAGS= LDFLAGS=
==> Downloading https://www.kernel.org/pub/software/scm/git/git-manpages-2.23.0.tar.xz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/6a8411d0ce8a37af7bcc0b586a2e4bc4f286f54bdfdcbc2683fe76ed72ac457b--git-manpages-2.23.0.tar.xz
==> Downloading https://www.kernel.org/pub/software/scm/git/git-htmldocs-2.23.0.tar.xz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/1d663eba035cf8f0470de76ca1532a69932cb7003300e62d670ed7879d6883d4--git-htmldocs-2.23.0.tar.xz
==> Downloading https://cpan.metacpan.org/authors/id/R/RJ/RJBS/Net-SMTP-SSL-1.04.tar.gz
Already downloaded: /Users/adam/Library/Caches/Homebrew/downloads/0726c68dac390b617cefcc880f076b50795463c1fba286d46a53082ebed2bc9d--Net-SMTP-SSL-1.04.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d

zsh completions and functions have been installed to:
  /usr/local/share/zsh/site-functions

Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/git
==> Summary
🍺  /usr/local/Cellar/git/HEAD-4c86140_1: 1,544 files, 45.0MB, built in 1 minute 27 seconds

✔ adam@wheatley ~
Wed 18 Sep 21:42 $ /usr/local/bin/git --version  # But look, the version number has changed
git version 2.23.0.rc2.292.g4c86140027

✔ adam@wheatley ~
Wed 18 Sep 21:43 $ cd ~/Library/Caches/Homebrew/git--git

✔ adam@wheatley ~/Library/Caches/Homebrew/git--git [master|✔]
Wed 18 Sep 21:44 $ git config --get remote.origin.tagOpt  # And here's why...
--no-tags
```

I don't think they're useful, but [here's the `brew gist-logs git` output](https://gist.github.com/me-and/ce7b8e82a34ddf2984dfde1882a80d34)

**Comment by SMillerDev** [comment]
As far as I know head installs don't update unless you remove and install them again. That's why they're discouraged unless you really need a feature that is unreleased.

**Comment by me-and** [comment]
Oh, okay. I wasn't aware of that limitation. 

Nonetheless, I think it's wrong that the first head install picks up tags, and subsequent reinstalls don't. At least in the context of the Git package, that means you get this inconsistent behaviour with how version numbers get reported.

**Comment by fxcoudert** [comment]
I think the `--fetch-HEAD` option is what you want. Otherwise, we would welcome a documentation fix, or maybe a feature request for `brew` itself: https://github.com/Homebrew/brew
